### PR TITLE
Truncate support decimal

### LINKED ISSUE #2605: TRUNCATE() result error
### Steps to reproduce the behavior (Required)
`SELECT TRUNCATE(0.0, 1326124691);`

### Expected behavior (Required)
0.000000000000000000000000000000
### Real behavior (Required)
```
+-------------------------+
| truncate(0, 1326124691) |
+-------------------------+
|                     nan |
+-------------------------+
```
### StarRocks version (Required)
 - You can get the StarRocks version by executing SQL `select current_version()`
```
select current_version();
+------------------------------------------+
| current_version()                        |
+------------------------------------------+
| PIPELINE_SQLANCER_MASTER_RELEASE 635482f |
+------------------------------------------+
```

### PR DESCRIPTION
**Pull request by liuyehcf**
## What type of PR is this：
- [ ] bug
- [ ] feature
- [x] enhancement
- [ ] others

## Which issues of this PR fixes ：

Fixes #2605

## Problem Summary(Required) ：

Related-pr [3744](https://github.com/StarRocks/starrocks/pull/3744)

Double version of truncate and decimal version of truncate share the same function name `truncate`

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 7 / 32 (21.88%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/analyzer/DecimalV3FunctionAnalyzer.java|0|20|00.00%|
|:large_blue_circle:|com/starrocks/catalog/Function.java|1|3|33.33%|
|:large_blue_circle:|com/starrocks/sql/analyzer/ExpressionAnalyzer.java|6|9|66.67%|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 31 / 34 (91.18%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/analyzer/ExpressionAnalyzer.java|7|10|70.00%|
|:large_blue_circle:|com/starrocks/sql/analyzer/DecimalV3FunctionAnalyzer.java|21|21|100.00%|
|:large_blue_circle:|com/starrocks/catalog/Function.java|3|3|100.00%|

File: gensrc/script/vectorized/vectorized_functions.py [hunk 1]
```diff
@@ -52,8 +52,8 @@
     [10131, "dround", "BIGINT", ["DOUBLE"], "MathFunctions::round"],
     [10132, "round", "DOUBLE", ["DOUBLE", "INT"], "MathFunctions::round_up_to"],
     [10133, "dround", "DOUBLE", ["DOUBLE", "INT"], "MathFunctions::round_up_to"],
-    [10134, "truncate", "DOUBLE", ["DOUBLE", "INT"], "MathFunctions::truncate"],
-    [10135, "truncate_decimal", "DECIMAL128", ["DECIMAL128", "INT"], "MathFunctions::truncate_decimal128"],
+    [10134, "truncate", "DECIMAL128", ["DECIMAL128", "INT"], "MathFunctions::truncate_decimal128"],
+    [10135, "truncate", "DOUBLE", ["DOUBLE", "INT"], "MathFunctions::truncate"],
```
Comment (Seaven) [inline gensrc/script/vectorized/vectorized_functions.py:56]: The modify will case compatibility problem when be was updated and fe didn't update

Comment (liuyehcf) [reply gensrc/script/vectorized/vectorized_functions.py:56]: Ok, let me figure it out

**Comment by Seaven** [comment]
You can try the way:  find decimal function by `Function.CompareMode.IS_IDENTICAL` mode when the type of first param is decimal

File: fe/fe-core/src/main/java/com/starrocks/sql/analyzer/ExpressionAnalyzer.java [hunk 2]
```diff
@@ -515,7 +515,7 @@ public Void visitFunctionCall(FunctionCallExpr node, Scope scope) {
                 }
             } else if (Arrays.stream(argumentTypes).anyMatch(Type::isDecimalV3) ||
                     // truncate have only decimal version
-                    FunctionSet.TRUNCATE_DECIMAL.equalsIgnoreCase(fnName)) {
+                    FunctionSet.TRUNCATE.equalsIgnoreCase(fnName)) {
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/sql/analyzer/ExpressionAnalyzer.java:518]: BYW, this is a useful check?

**Comment by liuyehcf** [comment]
> 

Good suggestion, thanks.

File: fe/fe-core/src/main/java/com/starrocks/sql/analyzer/ExpressionAnalyzer.java (hunk 2 above)
Comment (liuyehcf) [reply fe/fe-core/src/main/java/com/starrocks/sql/analyzer/ExpressionAnalyzer.java:518]: > You can try the way: find decimal function by `Function.CompareMode.IS_IDENTICAL` mode when the type of first param is decimal

Thanks for advice!

**Comment by liuyehcf** [comment]
> You can try the way: find decimal function by `Function.CompareMode.IS_IDENTICAL` mode when the type of first param is decimal

Thanks for advice

**Comment by liuyehcf** [comment]
> 

When enable `IS_NONSTRICT_SUPERTYPE_OF`, it will first go through `IS_IDENTICAL` mode, and cannot find a match. Then turn to search in `IS_INDISTINGUISHABLE` mode and `IS_SUPERTYPE_OF` mode by order. And we need to ensure that the search process can normally match the decimal version in the `IS_INDISTINGUISHABLE` stage, otherwise, always get double version in `IS_SUPERTYPE_OF` stage, because function id of double version is smaller than decimal's

**Comment by liuyehcf** [comment]
> You can try the way: find decimal function by `Function.CompareMode.IS_IDENTICAL` mode when the type of first param is decimal

When enable `IS_NONSTRICT_SUPERTYPE_OF`, it will first go through `IS_IDENTICAL` mode, and cannot find a match. Then turn to search in `IS_INDISTINGUISHABLE` mode and `IS_SUPERTYPE_OF` mode by order. And we need to ensure that the search process can normally match the decimal version in the `IS_INDISTINGUISHABLE` stage, otherwise, always get double version in `IS_SUPERTYPE_OF` stage, because function id of double version is smaller than decimal's

File: fe/fe-core/src/main/java/com/starrocks/sql/analyzer/ExpressionAnalyzer.java (hunk 2 above)
Comment (liuyehcf) [reply fe/fe-core/src/main/java/com/starrocks/sql/analyzer/ExpressionAnalyzer.java:518]: Let me check

Comment (liuyehcf) [reply fe/fe-core/src/main/java/com/starrocks/sql/analyzer/ExpressionAnalyzer.java]: It canbe removed

File: gensrc/script/vectorized/vectorized_functions.py (hunk 1 above)
Comment (liuyehcf) [reply gensrc/script/vectorized/vectorized_functions.py]: fixed

**Comment by liuyehcf** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 27 / 30 (90.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/analyzer/ExpressionAnalyzer.java|7|10|70.00%|
|:large_blue_circle:|com/starrocks/sql/analyzer/DecimalV3FunctionAnalyzer.java|17|17|100.00%|
|:large_blue_circle:|com/starrocks/catalog/Function.java|3|3|100.00%|

**Review by murphyatwork — DISMISSED** [review]

File: fe/fe-core/src/main/java/com/starrocks/catalog/Function.java [hunk 3]
```diff
@@ -442,6 +442,10 @@ private boolean isIndistinguishable(Function o) {
             if (o.argTypes[i].isNull() || this.argTypes[i].isNull()) {
                 continue;
             }
+            if (FunctionSet.TRUNCATE.equalsIgnoreCase(name.getFunction()) &&
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/catalog/Function.java:445]: This is seems very strange.

Comment (liuyehcf) [reply fe/fe-core/src/main/java/com/starrocks/catalog/Function.java:445]: Yes, otherwise it will always match double version, do you have any suggestion?

**Comment by Seaven** [comment]
> > You can try the way: find decimal function by `Function.CompareMode.IS_IDENTICAL` mode when the type of first param is decimal
> 
> When enable `IS_NONSTRICT_SUPERTYPE_OF`, it will first go through `IS_IDENTICAL` mode, and cannot find a match. Then turn to search in `IS_INDISTINGUISHABLE` mode and `IS_SUPERTYPE_OF` mode by order. And we need to ensure that the search process can normally match the decimal version in the `IS_INDISTINGUISHABLE` stage, otherwise, it will always return double version at `IS_SUPERTYPE_OF` stage, because function id of double version is smaller than decimal's

I see, I think the reason about can't match the decimal version is the second param can't match, it's really hard... Must keep the order of decimal version is before double version, or set function id to 10129?

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 26 / 29 (89.66%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/analyzer/ExpressionAnalyzer.java|7|10|70.00%|
|:large_blue_circle:|com/starrocks/sql/analyzer/DecimalV3FunctionAnalyzer.java|17|17|100.00%|
|:large_blue_circle:|com/starrocks/catalog/Function.java|2|2|100.00%|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 26 / 29 (89.66%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/analyzer/ExpressionAnalyzer.java|7|10|70.00%|
|:large_blue_circle:|com/starrocks/sql/analyzer/DecimalV3FunctionAnalyzer.java|17|17|100.00%|
|:large_blue_circle:|com/starrocks/catalog/Function.java|2|2|100.00%|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 26 / 29 (89.66%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/analyzer/ExpressionAnalyzer.java|7|10|70.00%|
|:large_blue_circle:|com/starrocks/sql/analyzer/DecimalV3FunctionAnalyzer.java|17|17|100.00%|
|:large_blue_circle:|com/starrocks/catalog/Function.java|2|2|100.00%|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 27 / 30 (90.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/analyzer/ExpressionAnalyzer.java|7|10|70.00%|
|:large_blue_circle:|com/starrocks/sql/analyzer/DecimalV3FunctionAnalyzer.java|18|18|100.00%|
|:large_blue_circle:|com/starrocks/catalog/Function.java|2|2|100.00%|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 27 / 30 (90.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/analyzer/ExpressionAnalyzer.java|7|10|70.00%|
|:large_blue_circle:|com/starrocks/sql/analyzer/DecimalV3FunctionAnalyzer.java|18|18|100.00%|
|:large_blue_circle:|com/starrocks/catalog/Function.java|2|2|100.00%|

**Comment by liuyehcf** [comment]
run starrocks_fe_unittest

**Comment by liuyehcf** [comment]
starrocks_fe_unittest

**Comment by liuyehcf** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 27 / 30 (90.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/analyzer/ExpressionAnalyzer.java|7|10|70.00%|
|:large_blue_circle:|com/starrocks/sql/analyzer/DecimalV3FunctionAnalyzer.java|18|18|100.00%|
|:large_blue_circle:|com/starrocks/catalog/Function.java|2|2|100.00%|

**Review by murphyatwork — APPROVED** [review]

**Review by Seaven — APPROVED** [review]

**Review by kangkaisen — APPROVED** [review]

**Pull request merged**